pkg/controllers/centralizedhpa/centralizedhpa_controller.go

```diff
@@ -0,0 +1,1293 @@
+
+// CentralizedHorizontalPodAutoscalerController is to sync HorizontalPodAutoscaler.
+type CentralizedHorizontalPodAutoscalerController struct {
+	client.Client
+	ReplicaCalc               *ReplicaCalculator
+	ClusterScaleClientSetFunc func(string, client.Client) (*util.ClusterScaleClient, error)
+	RESTMapper                meta.RESTMapper
+	EventRecorder             record.EventRecorder
+	TypedInformerManager      typedmanager.MultiClusterInformerManager
+
+	monitor monitor.Monitor
+
+	HorizontalPodAutoscalerSyncPeroid time.Duration
+	DownscaleStabilisationWindow      time.Duration
+	// Latest unstabilized recommendations for each autoscaler.
+	recommendations     map[string][]timestampedRecommendation
+	recommendationsLock sync.Mutex
+
+	// Latest autoscaler events
+	scaleUpEvents       map[string][]timestampedScaleEvent
+	scaleUpEventsLock   sync.RWMutex
+	scaleDownEvents     map[string][]timestampedScaleEvent
+	scaleDownEventsLock sync.RWMutex
+
+	// Storage of HPAs and their selectors.
+	hpaSelectors    *selectors.BiMultimap
+	hpaSelectorsMux sync.Mutex
+
+	RateLimiterOptions ratelimiterflag.Options
+}
+
+type timestampedScaleEvent struct {
+	replicaChange int32 // positive for scaleUp, negative for scaleDown
+	timestamp     time.Time
+	outdated      bool
+}
+
+type timestampedRecommendation struct {
+	recommendation int32
+	timestamp      time.Time
+}
+
+// SetupWithManager creates a controller and register to controller manager.
+func (c *CentralizedHorizontalPodAutoscalerController) SetupWithManager(mgr controllerruntime.Manager) error {
+	c.recommendations = map[string][]timestampedRecommendation{}
+	c.scaleUpEvents = map[string][]timestampedScaleEvent{}
+	c.scaleDownEvents = map[string][]timestampedScaleEvent{}
+	c.hpaSelectors = selectors.NewBiMultimap()
+	c.monitor = monitor.New()
+	return controllerruntime.NewControllerManagedBy(mgr).
+		For(&autoscalingv1alpha1.FederatedHPA{}).
+		WithOptions(controller.Options{RateLimiter: ratelimiterflag.DefaultControllerRateLimiter(c.RateLimiterOptions)}).
+		Complete(c)
+}
+
+// Reconcile performs a full reconciliation for the object referred to by the Request.
+// The Controller will requeue the Request to be processed again if an error is non-nil or
+// Result.Requeue is true, otherwise upon completion it will remove the work from the queue.
+func (c *CentralizedHorizontalPodAutoscalerController) Reconcile(ctx context.Context, req controllerruntime.Request) (controllerruntime.Result, error) {
+	klog.V(4).Infof("Reconciling HorizontalPodAutoscaler %s.", req.NamespacedName.String())
+
+	hpa := &autoscalingv1alpha1.FederatedHPA{}
+	key := req.NamespacedName.String()
+	if err := c.Client.Get(ctx, req.NamespacedName, hpa); err != nil {
+		if apierrors.IsNotFound(err) {
+			klog.Infof("Horizontal Pod Autoscaler %s has been deleted in %s", req.Name, req.Namespace)
+			c.recommendationsLock.Lock()
+			delete(c.recommendations, key)
+			c.recommendationsLock.Unlock()
+
+			c.scaleUpEventsLock.Lock()
```

Comment:
I see there is no lock in the original k8s implementation, is there any reason?

Reply:
Lock is added in the recent release, e.g. 1.27. The main reason is that in the older version, there will be only one worker so there will not sync-write issue in map.